subsys/bluetooth/rpc/common/serialize.c

```diff
@@ -4,596 +4,409 @@
 {
-	CborError err;
-
-	if (is_encoder_invalid(encoder)) {
-		return;
-	}
-
-	err = cbor_encode_null(encoder);
-
-	if (err != CborNoError) {
-		set_encoder_invalid(encoder, err);
-	}
+	zcbor_nil_put(ctx->zs, NULL);
 }
 
-void ser_encode_undefined(CborEncoder *encoder)
+void ser_encode_undefined(struct nrf_rpc_cbor_ctx *ctx)
 {
-	CborError err;
-
-	if (is_encoder_invalid(encoder)) {
-		return;
-	}
-
-	err = cbor_encode_undefined(encoder);
-	if (err != CborNoError) {
-		set_encoder_invalid(encoder, err);
-	}
+	zcbor_undefined_put(ctx->zs, NULL);
 }
 
-void ser_encode_bool(CborEncoder *encoder, bool value)
+void ser_encode_bool(struct nrf_rpc_cbor_ctx *ctx, bool value)
 {
-	CborError err;
-
-	if (is_encoder_invalid(encoder)) {
-		return;
-	}
-
-	err = cbor_encode_boolean(encoder, value);
-	if (err != CborNoError) {
-		set_encoder_invalid(encoder, err);
-	}
+	zcbor_bool_put(ctx->zs, value);
 }
 
-void ser_encode_uint(CborEncoder *encoder, uint32_t value)
+void ser_encode_uint(struct nrf_rpc_cbor_ctx *ctx, uint32_t value)
 {
-	CborError err;
-
-	if (is_encoder_invalid(encoder)) {
-		return;
-	}
-
-	err = cbor_encode_uint(encoder, value);
-	if (err != CborNoError) {
-		set_encoder_invalid(encoder, err);
-	}
+	zcbor_uint32_put(ctx->zs, value);
 }
 
-
-void ser_encode_int(CborEncoder *encoder, int32_t value)
+void ser_encode_int(struct nrf_rpc_cbor_ctx *ctx, int32_t value)
 {
-	CborError err;
-
-	if (is_encoder_invalid(encoder)) {
-		return;
-	}
-
-	err = cbor_encode_int(encoder, value);
-	if (err != CborNoError) {
-		set_encoder_invalid(encoder, err);
-	}
+	zcbor_int32_put(ctx->zs, value);
 }
 
-void ser_encode_str(CborEncoder *encoder, const char *value, int len)
+void ser_encode_str(struct nrf_rpc_cbor_ctx *ctx, const char *value, int len)
 {
-	CborError err;
-
-	if (is_encoder_invalid(encoder)) {
-		return;
-	}
-
 	if (!value) {
-		err = cbor_encode_null(encoder);
-	} else if (len < 0) {
-		err = cbor_encode_text_stringz(encoder, value);
+		zcbor_nil_put(ctx->zs, NULL);
 	} else {
-		err = cbor_encode_text_string(encoder, value, len);
+		zcbor_tstr_encode_ptr(ctx->zs, value, len);
 	}
-
-	if (err != CborNoError) {
-		set_encoder_invalid(encoder, err);
-	}
-
 }
 
-void ser_encode_buffer(CborEncoder *encoder, const void *data, size_t size)
+void ser_encode_buffer(struct nrf_rpc_cbor_ctx *ctx, const void *data, size_t size)
 {
-	CborError err;
-
-	if (is_encoder_invalid(encoder)) {
-		return;
-	}
-
 	if (!data) {
-		err = cbor_encode_null(encoder);
+		zcbor_nil_put(ctx->zs, NULL);
 	} else {
-		err = cbor_encode_byte_string(encoder, (const uint8_t *)data, size);
-	}
-
-	if (err != CborNoError) {
-		set_encoder_invalid(encoder, err);
+		zcbor_bstr_encode_ptr(ctx->zs, data, size);
 	}
 }
 
-void ser_encode_callback(CborEncoder *encoder, void *callback)
+void ser_encode_callback(struct nrf_rpc_cbor_ctx *ctx, void *callback)
 {
-	CborError err;
 	int slot;
 
-	if (is_encoder_invalid(encoder)) {
+	if (is_encoder_invalid(ctx)) {
```

Comment:
We can skip that. Allocating slot for `callback` if encoder is in invalid state is not a problem. It will be simply unused. It is very unusual situation, so this unused slot will not cause any problem.

Reply:
But if we don't put too much changes, we can optimize the code later.